openedx/adg/lms/applications/analytics.py

```diff
@@ -0,0 +1,191 @@
+
+
+def _get_start_date_and_end_date(position):
+    """
+    Check if the position is ongoing (current), and returns the start_date and the end_date of the
+    position.
+
+    Arguments:
+        position (UserStartAndEndDates): WorkExperience or Education to get the start and end date for
+
+    Returns:
+        (str, str): A tuple containing start_date and end_date in a tuple. If the position is ongoing
+            the end_date is `In Progress`.
+    """
+    format_for_month_and_year_only = '%B, %G'
+    start_date = get_formatted_date(
+        datetime(position.date_started_year, position.date_started_month, 1), format_for_month_and_year_only
+    )
+
+    is_position_current = (
+        (hasattr(position, 'is_current_position') and position.is_current_position) or
+        (hasattr(position, 'is_in_progress') and position.is_in_progress)
+    )
+    if is_position_current:
+        end_date = IN_PROGRESS
+    else:
+        end_date = get_formatted_date(
+            datetime(position.date_completed_year, position.date_completed_month, 1), format_for_month_and_year_only
+        )
+
+    return start_date, end_date
+
+
+def emit_enrollment_event_for_bu_courses(user, status):
+    """
+    Emit event to update the progress of Business Unit courses
+
+    Arguments:
+        user (User): User for which emit the BU courses progress update
+        status (str): Progress of the bu courses
+    """
+    segment.identify(user.id, {'businessLineCoursesStatus': status})
+
+
+def emit_course_progress_event(user, course, grade=None):
+    """
+    Emit the progress of a course when a course is Started, Completed or Failed. Emits
+    an event for omni courses if a course is started, or passed
+
+    Arguments:
+        user (User): User associated with the course
+        course (CourseOverview): The course for which we are emitting the progress
+        grade (PersistentCourseGrade): The grade for the given course and user
+    """
+    from openedx.adg.lms.applications.helpers import has_attempted_all_modules
+
+    course_language_name = get_language_info(course.language).get('name', course.language)
+    course_properties = {'courseName': course.id.course, 'courseLanguage': course_language_name}
+
+    if grade:
+        course_properties['letterGrade'] = text_type(grade.letter_grade)
+        course_properties['percentageGrade'] = convert_float_point_to_percentage(grade.percent_grade)
+
+        if grade.passed_timestamp and grade.letter_grade:
+            course_event = COURSE_COMPLETED
+            emit_enrollment_event_for_omni_courses(user.id, course, COURSE_COMPLETED)
+        else:
+            course_event = COURSE_FAILED if has_attempted_all_modules(user, course) else None
+    else:
+        course_event = COURSE_STARTED
+        emit_enrollment_event_for_omni_courses(user.id, course, COURSE_STARTED)
+
+    if course_event in COURSE_EVENTS:
+        segment.track(user.id, COURSE_EVENTS[course_event], course_properties)
```

Comment:
Why do we need this. According to this function, under all circumstances, the value in `course_event` would either be COURSE_COMPLETED, COURSE_FAILED, COURSE_STARTED or None. Can't we simply replace these two lines with the following:

if course_event:
     segment.track(user.id, course_event, course_properties)

Reply:
I wrote it that way so it is safe for future extensions. But you make a good point, I can refactor this.